src/librustc/traits/mod.rs

```diff
@@ -193,6 +193,8 @@ pub enum ObligationCauseCode<'tcx> {
 
     ImplDerivedObligation(DerivedObligationCause<'tcx>),
 
+    DerivedCauseCode(Box<ObligationCauseCode<'tcx>>),
```

Comment:
Don't we have a system for this already?

Reply:
The closest is what is above but that requires a predicates to be available always, and that wasn't feasible for the cases I was handling.

Reply:
WF definitely has WF predicates, so at least that should be handleable.

I'll take a second look at the diff and note the places `DerivedCauseCode` is created in.

Reply:
Okay, I see, `DerivedObligationCause` is a bit limited, you want to replace the `TraitRef` with a `Predicate` (which would let you put a WF predicate in there).

But still, I'd prefer if this side of the PR was done separately from the `ItemObligation` side.